tools/util/ReplayOperationManager.hpp

```diff
@@ -0,0 +1,1155 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#ifndef REPLAY_ReplayOperationManager_HPP
+#define REPLAY_ReplayOperationManager_HPP
+
+#include <iostream>
+#include <cstdint>
+#include <vector>
+
+#include "umpire/Allocator.hpp"
+#include "umpire/strategy/AllocationAdvisor.hpp"
+#include "umpire/strategy/SizeLimiter.hpp"
+#include "umpire/strategy/MixedPool.hpp"
+#include "umpire/strategy/MonotonicAllocationStrategy.hpp"
+#include "umpire/strategy/SlotPool.hpp"
+#include "umpire/strategy/ThreadSafeAllocator.hpp"
+#include "umpire/ResourceManager.hpp"
+
+class ReplayOperation {
+public:
+ using AllocationOpMap = std::unordered_map<uint64_t, ReplayOperation*>;
+  std::function<void ()> op;
+
+  ReplayOperation(
+      std::vector<umpire::Allocator>& alloc_array,
+      AllocationOpMap& alloc_operations
+  ) :   m_alloc_array(alloc_array)
+      , m_alloc_operations(alloc_operations)
+  {
+  }
+
+  void run()
+  {
+    op();
+  }
+
+  //
+  // AllocationAdvisor
+  //
+  template <typename... Args>
+  void makeAdvisor(
+      const bool introspection,
+      const std::string& allocator_name,
+      const std::string& base_allocator_name,
+      Args&&... args
+  )
+  {
+    if (introspection) {
+      op = [=]() {
+        auto& rm = umpire::ResourceManager::getInstance();
+
+        rm.makeAllocator<umpire::strategy::AllocationAdvisor, true>
+          ( allocator_name, rm.getAllocator(base_allocator_name),
+            std::forward<Args>(args)...);
+
+        this->m_alloc_array.push_back(
+            umpire::Allocator(rm.getAllocator(allocator_name)));
```

Comment:
```suggestion
            rm.getAllocator(allocator_name));
```

I think you were doing an extra copy-construction here.

Reply:
Fixed